### PR TITLE
feat!: add websocket interface for pubsub

### LINKED ISSUE #1668: [Feature] Create websocket manager
- Create websocket manager for our pubsub
- It acts as an interface between clients and the pubsub system.
- The protocol makes it possible for messages to be published and subscribed.

### PR DESCRIPTION
**Pull request by shubham4443**
Fixes #1668
The PR adds the ability to perform publish/subscribe operations with the websocket protocol.

### CODE REVIEW

File: sc-poc/modules/pubsub/websocket.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+	"github.com/spacecloud-io/space-cloud/managers/apis"
+)
+
+// TODO: channels may or may not have prefix slash
+
+// GetRoutes returns all the apis that are exposed by this app
+func (a *App) GetAPIRoutes() apis.APIs {
+	channels := a.Channels()
+	apis := []*apis.API{}
```
Comment (YourTechBud) [inline sc-poc/modules/pubsub/websocket.go:23]: I suggest we make this api in `Provision` and simply return it here

File: sc-poc/modules/pubsub/websocket.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+			// Create the websocket connection
+			conn, _, _, err := ws.UpgradeHTTP(r, w)
+			if err != nil {
+				a.logger.Error("could not establish websocket connection", zap.String("channel", channelName), zap.Error(err))
+				return
+			}
+
+			// Go routine to receive messages from the websocket connection and publish
+			// them into the watermill channel
+			go func() {
```
Comment (YourTechBud) [inline sc-poc/modules/pubsub/websocket.go:51]: I don't think we need to run this as a seperate goroutine.

You can confirm this by starting a subscriber then trying to reload the config. If the config gets reloaded without affecting the socket we are good.

File: sc-poc/modules/pubsub/websocket.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+					if message.Event == messageEvent {
+						var pubMsg PublishMessage
+						err = mapstructure.Decode(message.Data, &pubMsg)
+						if err != nil {
+							a.logger.Error("could not decode data", zap.String("channel", channelName), zap.Error(err))
+							continue
+						}
+
+						if pubMsg.ID == "" {
+							pubMsg.ID = uuid.NewString()
```
Comment (YourTechBud) [inline sc-poc/modules/pubsub/websocket.go:77]: Use ksuid

File: sc-poc/modules/pubsub/websocket.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+			// Create the websocket connection
+			conn, _, _, err := ws.UpgradeHTTP(r, w)
+			if err != nil {
+				a.logger.Error("could not establish websocket connection", zap.String("channel", channelName), zap.Error(err))
+				return
+			}
+
+			// Wait for subscribe event
+			for {
```
Comment (YourTechBud) [inline sc-poc/modules/pubsub/websocket.go:108]: This implementation won't work cause it doesn't leave room for you to check for the unsubscribe event that could come in next

**Review by YourTechBud — CHANGES_REQUESTED** [review]

File: sc-poc/modules/pubsub/websocket.go (hunk 4 above)
Comment (shubham4443) [reply sc-poc/modules/pubsub/websocket.go:108]: Should we use a variable which stores the state(subscribed/not subscribed) in which the client is in right now?

File: sc-poc/modules/pubsub/types.go [hunk 5]
```diff
@@ -1,5 +1,21 @@
 package pubsub
```
Comment (YourTechBud) [inline sc-poc/modules/pubsub/types.go:2]: Export `eventType` and the various constants

**Review by YourTechBud — CHANGES_REQUESTED** [review]

**Pull request merged**